rero_ils/modules/templates/schemas/json.py

```diff
@@ -0,0 +1,109 @@
+
+schema_from_template = partial(schema_from_context, schema=Template.schema)
+
+
+class TemplateMetadataSchemaV1(StrictKeysMixin):
+    """Marshmallow schema for the `Template` metadata."""
+
+    pid = SanitizedUnicode()
+    template_type = SanitizedUnicode(required=True)
+    name = SanitizedUnicode(required=True)
+    description = SanitizedUnicode()
+    visibility = SanitizedUnicode(
+        required=True,
+        validate=OneOf([TemplateVisibility.PUBLIC, TemplateVisibility.PRIVATE])
+    )
+    data = fields.Dict()
+    creator = fields.Dict()
+    organisation = fields.Dict()
+    schema = GenFunction(
+        load_only=True,
+        attribute="$schema",
+        data_key="$schema",
+        deserialize=schema_from_template
+    )
+
+    @validates('visibility')
+    @http_applicable_method('POST')
+    def validate_visibility(self, data, **kwargs):
+        """Validate the visibility field through REST API request.
+
+        Through the API, a template could only be created with `private`
+        visibility. To set a template in `public` visibility, an authorized
+        user must update an existing template.
+
+        :param data: the visibility field value.
+        :param kwargs: any additional named arguments.
+        :raises ValidationError: if error has detected on visibility attribute
+        """
+        if data == TemplateVisibility.PUBLIC:
+            raise ValidationError(
+                'Template can be created only with `private` visibility')
+
+    @validates_schema()
+    @http_applicable_method('PUT')
+    def validate_visibility_changes(self, data, **kwargs):
+        """Validate `visibility` changes through REST API request.
+
+        Updates on `visibility` attribute is subject to restriction depending
+        on current connected user ; only FULL_PERMISSIONS and
+        LIBRARY_ADMINISTRATOR users could change `visibility` attribute.
+
+        :param data: the json data to validate.
+        :param kwargs: additional named arguments.
+        :raises ValidationError: if error has detected on visibility attribute
+        """
+        # Load DB record
+        db_record = Template.get_record_by_pid(data.get('pid'))
+        if not db_record:
+            raise ValidationError(f'Unable to load Template#{data.get("pid")}')
+
+        # Check if visibility of the template changed. If not, we can stop
+        # the validation process.
+        if db_record.get('visibility') == data.get('visibility'):
+            return
+
+        # Only lib_admin and full_permission roles can change visibility field
+        allowed_roles = [
+            UserRole.FULL_PERMISSIONS,
+            UserRole.LIBRARY_ADMINISTRATOR
+        ]
+        user_roles = set()
+        if current_librarian:
+            user_roles = set(current_librarian.get('roles', []))
+        if not user_roles.intersection(allowed_roles):
+            error_message = 'You are not allowed to change template visibility'
```

Comment:
Is it really the role of the marshmallow or it should be done at the permission level?

Reply:
This is all the problem and the reason why I use marshmallow.
This kind of changes should only be controlled through API requests, using console/script we want allow any changes.
But permission layer applies both via API and via console. Using marshmallow, I can control changes only at API level using `@validates` or `@validates_schema`